src/Mono.Android/Android.Runtime/JNIEnv.cs

```diff
@@ -251,9 +249,15 @@ static void ManualJavaObjectDispose (Java.Lang.Object obj)
 
 		internal static void PropagateUncaughtException (IntPtr env, IntPtr javaThreadPtr, IntPtr javaExceptionPtr)
```

Comment:
Maybe it would make more sense to just leave this entirely in Java & C...?

Reply:
We should *absolutely* move all this stuff into `XamarinUncaughtExceptionHandler`.

Here's my conceptual problem: `Thread.getDefaultUncaughtExceptionHandler()` forms a "linked list", and it is the responsibility of all `Thread.UncaughtExceptionHandler` implementations to propagate the exception to their "parent" handler within `Thread.UncaughtExceptionHandler.uncaughtException()`.

The problem here is twofold:

 1. `UncaughtExceptionHandler` instantiation is "responsible" for calling `Thread.DefaultUncaughtExceptionHandler` and obtaining the "parent" handler, and
 2. `UncaughtExceptionHandler` is only instantiated *when there's an unhandled exception*, so it's not "clean" as of process startup, and
 3. Propagating the unhandled exception is thus poorly defined, plus
 4. It's *way* confusion trying to keep these four different spots in mind and in sync (`XamarinUncaughtExceptionHandler`, `UncaughtExceptionHandler`, `JNIEnv.PropagateUncaughtException()`, `UncaughtExceptionHandler.UncaughtException()`).

It does not give me a good feeling.

Instead, let's "move" the "default uncaught exception handler linked list management" of `UncaughtExceptionHandler` into `XamarinUncaughtExceptionHandler`, and then move `UncaughtExceptionHandler.UncaughtException()` into `JNIEnv.PropagateUncaughtException()`.

This should kill two places to try to keep track of things, simplifying logic.

Reply:
Makes sense, changes applied. Working on this I wanted to make the minimum amount of changes to minimize potential of breaking something that's been in place for years, but I agree with your assessment 100%